### PR TITLE
Add TPM feature page

### PR DESCRIPTION
**Pull request by mz-pdm**
Changes proposed in this pull request:

- Initial version of TPM feature page. It's WIP, will be updated as needed.

I confirm that this pull request was submitted according to the [contribution guidelines](https://github.com/oVirt/ovirt-site/blob/master/CONTRIBUTING.md): @mz-pdm 

This pull request needs review by: @ahadas

### CODE REVIEW

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [inline source/develop/release-management/features/virt/tpm-device.md:61]: I'm not sure the approach with VM.destroy is reliable enough itself even when not considering other problems. Considering that TPM data doesn't change that often under normal circumstances, it could be complemented with stats reporting. A periodic job would watch for TPM data changes and if the data changed & remained the same in the next cycle then it would be reported as part of VM stats. I'd need more knowledge about how things work in Engine to implement that correctly but I think the basic idea could work. What do you think?

File: source/develop/release-management/features/virt/tpm-device.md
Comment (sandrobonazzola) [inline source/develop/release-management/features/virt/tpm-device.md:70]: Can you please add also a section on how to test the feature?

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [inline source/develop/release-management/features/virt/tpm-device.md:68]: That's an interesting approach.
Maybe we can combine the two together? I mean, to let VDSM saves the data to those volumes and read from them when starting the VMs instead of sending the data to the engine?

Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:68]: I discussed the idea of putting the data to storage with Nir and it doesn't look like a very simple thing. A volume would be needed with all the related arrangements and it needn't be absolutely reliable too -- we know storage problems can occur for many reasons. We came to the conclusion that a better idea could be to make a special Vdsm API call for retrieving TPM data that would be called periodically. Engine could ask for the data with a relatively long interval or whenever TPM data checksum reported by Vdsm changes (what would be the complications you mentioned above with storing VM data hashes on the Engine side and acting when they change?). The last call would be performed just before calling the final Destroy (to avoid problems with failures after the data is wiped on the host).
This way, Engine would be managing the data transfers, based on some information from Vdsm, which looks more robust than other solutions. Are there any obstacles on the Engine side to implement something like that?

Comment (ahadas) [reply source/develop/release-management/features/virt/tpm-device.md:68]: I think that storing this data (in vm_dynamic) and sending it on RunVm should be fairly simple. The complication I thought of was more about a mechanism to get the hash and detecting a change. We do that for the 'devices hash' - we have VmDevicesMonitoring that on engine startup loads the hash of each VM from the database and then stores that in-memory and compares it with the one we get from the stats call. It's not rocket science to introduce another hash that this module or an additional module would handle, but it can be simpler to use the same 'devices hash' for that as well if possible (that hash already covers some of the guest agent data afair).

What I like about storing this data on the storage though is that it sounds more aligned with our desire to be able to (re)start VMs without the engine being involved. I guess we wouldn't want to store such sensitive data in the metadata, right?

Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:68]: > What I like about storing this data on the storage though is that it sounds more aligned with our desire to be able to (re)start VMs without the engine being involved. 

That's a good remark and an argument important enough to consider using storage.

> I guess we wouldn't want to store such sensitive data in the metadata, right?

The data is already accessible to anybody with root access to the host and also, in most of the discussed implementation scenarios, to Vdsm. So I think having them in metadata is no big additional risk. But the question is whether we want to store larger data in metadata. While TPM capacity is AFAIK very limited (I'm not sure about emulated TPM though), it's still going to be at least 4 KB of encoded data with swtpm.

Still, using metadata would simplify things a lot. When using storage, we would have to arrange the volume, to ensure data can't be left in inconsistent state there (e.g. when the storage gets suddenly disconnected while TPM data is being written), and to deliver (consistent) data there, either by arranging swtpm to use the storage directly or making Vdsm copy it there.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md]: Done.

**Comment by mz-pdm** [comment]
The document has been reworked, section on testing was added, text on data security was separated and expanded, the implementation section was completely rewritten and expanded, and some more changes have been made. Four implementation approaches are described, we need to decide which one to attempt.

**Comment by sandrobonazzola** [comment]
Waiting for @ahadas approval before merging.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [inline source/develop/release-management/features/virt/tpm-device.md:37]: Would we need to block operations like export/clone of running VMs with TCP devices (those operations take a snapshot without memory, only of the disks, and correlate it with the VM configuration at the time of the snapshot)?

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [inline source/develop/release-management/features/virt/tpm-device.md:139]: Another drawback is that the size of these volume would likely be at lease 1G on block storage (like volumes that store vm configuration for snapshots with memory). This appears to be a reoccurring issue that also came up in the context of Q35 features..

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:71]: How live migration is handled? how it does not require anything on ovirt side?

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:76]: First 2 looks the same, data sent from engine when starting a vm.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:88]: We cannot support this. Engine assistance is always required to run vms, for example for attaching storage to host.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:89]: We depend on swtmp here. If it does not ensure consistency we cannot ensure this.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:94]: events are not good way to pass the data, but they can be used for signaling changes. Getting the data should be via new API for this purpose, or adding to existing API that returns information about a vm.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:96]: This does not make sense, requires creating special volume per vm. If we do this then there is no need to keep tpm data elsewhere, it will simply be in shared storage. This requires support from swtpm, they should be able to use block device for the data, and use direct I/O to ensure consistency.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:97]: Yes, this is a valid option, but this means adding new disk for tpm and handling the extra disk in all flows copying vm disks, like live storage migration. This is a lot of work.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:95]: This looks like abuse of the metadata, and also not reliabe as libvirt API can block for serveal reason.  I would not use it for anything except vm metadata.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:111]: This is not so simple, since the data may change again after we detected a change, but it can work.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:119]: We cannot start a vm on same host without enigne since vm xml and metadata is deleted when stooping a vm.

I think the only way that will work is to have new API called by engine to fetch tpm data, and 
split destroying a vm to 2 steps - stop the vm and undefine the vm. This way the vm data is not
lost.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:125]: But vm metadata is lost when you stop a vm, I don't see how it helps.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [reply source/develop/release-management/features/virt/tpm-device.md:139]: No, we can create 128m disks.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:143]: This approach makes sense only if swtpm writes to shared storage. If vdsm has to read and write this will only complicate things.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:147]: In this case libivrt will have to accept a path to tmp data.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:149]: This is not an issue if swtpm is written properly, ensuring atomic updates.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:150]: deleting on failure is not valid approach. 

The only way that will work is copy-on-write:

- Keep metadata sector pointing to the current data area
- keep 2 data areas

For each write, write to non-current data area. When the write is done
(fsync() succeeded), update the metadata sector to to point to this area.

On failure, the metadata sector still points to previous data area, no change
was done, and there nothing to cleanup.

On success, the metadata sector points to the new data.

If swtmp does not have something like this, there is no point to let it write
to shared storage since it will create inconsistent data.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:151]: Of course we can, this is exactly what we should do - move the complexity down to the platform.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (nirs) [inline source/develop/release-management/features/virt/tpm-device.md:152]: We can ask libivrt to support external tpm data, so libvirt does not create or remove anything. This way we have
complete control over the data, and we don't need to change libvirt flows.

So we can do:

1. Store tpm data on the host tpm store
2. Start vm with tpm store path
3. Stop vm (undefine, etc.)
4. Pass current tpm data to engine

We use the same method for other stuff like scratch disks during backup. Libvirt supports EXTERNAL flags to let the user control something that libvirt cannot mange. This sounds like the best approach.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:37]: Did you mean "with TPM devices"? That's a good question, I'm afraid we can fall into trouble with all kinds of disk snapshots when the VM is running. A possible solution could be checking TPM data before starting the snapshot and before finalizing it and if it doesn't match then to let the snapshot fail.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:71]: Live migration is handled transparently by libvirt. The only trouble with it could be if swtpm was arranged to write to the same shared storage volume directly on both the source and the destination.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:76]: Yes, but not handled the same way later.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:88]: If the storage is already attached (on the same host where the VM was running last time) then it may be possible. It's a completely different topic though, all we need here now is not to close the door to the functionality with TPM handling.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:89]: We may be forced to depend on some heuristics, unless upstream is willing to implement or accept improvements, such as if the local data doesn't change for a sufficiently long time interval then it can be considered OK.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:147]: Did you mean "tpm"? Yes, adding such a feature would be an option, but probably only with other arrangements such as prevention of data removal by libvirt.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:149]: Unfortunately it's not and we should explore whether that could be changed (in a sufficiently close future).

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:150]: Yes.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:151]: Ideally yes, but it may be a long run to fix upstream and get the fixes in downstream and it may not guarantee things won't break again in the future. Not that easy, we may still attempt that but other options for interim solutions must be also considered at the moment.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:152]: This would be necessary only when swtpm was writing to the shared storage directly. I like that option the least of all. In the other cases it makes sense that libvirt removes related data when we undefine a domain -- it's up to us to delay cleanup if we need that.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:89]: ... and is non-empty etc.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [reply source/develop/release-management/features/virt/tpm-device.md:139]: Ack, so it seems that we converge here to not store the TPM data on the storage domain but it could be useful to create smaller disks for snapshot's metadata volumes - allocating 1g as we do today for xml and some other vdsm properties is a waste

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [reply source/develop/release-management/features/virt/tpm-device.md:95]: right, that's a bit risky. so another option that we discussed off-line is to treat the TPM data as yet another data which does not change frequently, like the VM devices. so the computation of the reported hash will also take the TPM data into account and when the engine calls `dump` it will not only get the domain xml but also the TPM data as another in the dictionary. this mechanism may also possibly be used for reporting the secure boot's data later on.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (sandrobonazzola) [inline source/develop/release-management/features/virt/tpm-device.md:43]: please use <https://bugzilla.redhat.com/1855367> so it will also be available as link

Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md]: Done.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (mz-pdm) [reply source/develop/release-management/features/virt/tpm-device.md:95]: I added it as another option to the document.

File: source/develop/release-management/features/virt/tpm-device.md
Comment (ahadas) [reply source/develop/release-management/features/virt/tpm-device.md:51]: so let's remove that part?

**Review by ahadas — COMMENTED** [review]
from my side, let's just remove the cluster-level setting part and get this in

**Comment by mz-pdm** [comment]
Replaced TPM cluster setting with a per-VM setting.

**Review by ahadas — APPROVED** [review]

**Comment by sandrobonazzola** [comment]
There are a few conversation not resolved and a pending review request for @nirs . Can you please confirm if we can merge this or if we need further discussion?

**Review by mz-pdm — COMMENTED** [review]
Minor changes:

- Removed unavailability on POWER (it should be available in 8.3).
- Upstream swtpm fix mentioned.
- Cluster->VM option in Testing section.

I believe the document describes the current state of our thoughts. We can and will change it later as needed.

**Comment by ahadas** [comment]
> I believe the document describes the current state of our thoughts. We can and will change it later as needed.

+1
lgtm

**Comment by sandrobonazzola** [comment]
Merging based on @ahadas approval.

**Pull request merged**
